### PR TITLE
Comoment estimation

### PR DESCRIPTION
**Pull request by cdries**
This branch improves internal handling of the coskewness and cokurtosis matrices in the ES and VaR functions. The sample coskewness and cokurtosis estimators M3.MM and M4.MM are written in C exploiting the symmetric structure, leading to a speed improvement.
In addition, structured, shrinkage and EWMA estimators are added. The documentation is written to accommodate these changes and a vignette is added explaining all the new comoment estimators.

### CODE REVIEW

File: src/comomentsEstimators.c [hunk 1]
```diff
@@ -0,0 +1,2775 @@
...
+  double *rM3 = REAL(M3);
+  
+  int iter = 0;
+  // loop over the unique indices i <= j <= k
+  for (int ii = 0; ii < P; ii++) {
+    for (int jj = ii; jj < P; jj++) {
+      for (int kk = jj; kk < P; kk++) {
+        // compute coskewness element
+        double elem = 0.0;
+        if ((ii == jj) & (jj == kk)) elem = margskews[ii];
```
Comment (joshuaulrich) [inline src/comomentsEstimators.c]: This should be `&&`, not the bitwise operator `&`.  The latter is likely to cause future bugs.

File: src/comomentsEstimators.c [hunk 2]
```diff
@@ -0,0 +1,2775 @@
...
+  SEXP M3 = PROTECT(allocVector(REALSXP, P * (P + 1) * (P + 2) / 6));
+  double *rM3 = REAL(M3);
+  
+  int iter = 0;
+  // loop over the unique indices i <= j <= k
+  for (int ii = 0; ii < P; ii++) {
+    for (int jj = ii; jj < P; jj++) {
+      for (int kk = jj; kk < P; kk++) {
+        // compute and fill coskewness element
+        if ((ii == jj) & (jj == kk)) {
```
Comment (joshuaulrich) [inline src/comomentsEstimators.c]: Should be `&&`.

File: src/comomentsEstimators.c [hunk 3]
```diff
@@ -0,0 +1,2775 @@
...
+  int iter = 0;
+  // loop over the unique indices i <= j <= k <= l
+  for (int ii = 0; ii < P; ii++) {
+    for (int jj = ii; jj < P; jj++) {
+      for (int kk = jj; kk < P; kk++) {
+        for (int ll = kk; ll < P; ll++) {
+          // compute coskewness element
+          double elem = 0.0;
+          if (((ii == jj) & (jj == kk)) & (kk == ll)) elem = margk_iiii[ii];
+          if (((ii == jj) & (kk == ll)) & (jj != kk)) elem = margk_iikk[ii] * margk_iikk[kk];
```
Comment (joshuaulrich) [inline src/comomentsEstimators.c]: `&&` instead of `&`

File: R/MultivariateMoments.R [hunk 4]
```diff
@@ -17,185 +17,1522 @@
...
+    if (NCOL(f) == 1) {
+      beta <- apply(Xc, 2, function(a) cov(a, f) / var(f))
+      fc <- f - mean(f)
+      fvar <- mean(fc^2)
+      T2 <- fvar * beta %*% t(beta)
+      diag(T2) <- margvars
+    } else {
+      mod <- stats::lm(X ~ f)
+      beta <- t(mod$coefficients[-1,])
+      fcov <- cov(f) * (n - 1) / n
```
Comment (joshuaulrich) [inline R/MultivariateMoments.R]: `n` is not defined in this scope.  Is it supposed to be an argument to `M2.struct`?  Is it a function of other inputs?

**Comment by cdries** [comment]
Thanks for the feedback. I committed all changes and also checked if the same changes had to be made to other functions or files I worked on. This was the case for an 'F' to be replaced by FALSE in VaR.R, as you noted in the ES.R file.

The largest change was done in the functions `M2.shrink`, `M3.shrink` and `M4.shrink` where I changed the default values for `targets`, which was a vector of booleans, to accept a vector of integers selecting the targets. Hence making the long default argument redundant. The change is also addressed in the documentation.

**Comment by joshuaulrich** [comment]
Thanks for the quick response Dries!  I have another comment/suggestion, but it's not a blocker to merging this PR.  As I was looking through your code, my intuition was that many functions had similar code that could be extracted into smaller functions and re-used.  I don't have a specific example, but I will try to come up with one.

Another question: how did you test to ensure the output did not change between your refactored functions and those they replaced?

**Comment by cdries** [comment]
It is true that many functions have similar code since the methods for the different order moments are each time based upon the same concept, but the specific formulas (mainly the ones in the C files) to be used are different. I tried using smaller helper functions where possible and further suggestions are definitely welcome.

The only visible functions that are changed are `M3.MM`, `M4.MM`, `ES` and `VaR`. I tested those by comparing with the code in the master branch in a series of random inputs, trying different combinations of default and user settings. The differences were always within rounding error (1e-16).

The shrinkage `Mx.shrink` and structured `Mx.struct` estimators were tested for consistency by sampling from the model in increasing sample sizes and checking convergence.

I can write a test file for the functions and attach it here by tomorrow.

**Comment by cdries** [comment]
Attached is an R script combining the test files I used for checking the output between the branches. I ran it for different input variables, to be specified at the beginning of the script. The script then loads a branch, saves the output of all functions that have been changed and checks it using an absolute error metric to the output of the original branch.

I also ran `VaR` and `ES` for several combinations of input variables, but since I did not change the code of these functions the result is always the same. The internal functions are checked and produce the same output, these hidden functions are also tested in the attached R file.

If you need additional tests or more comments in the script, please let me know.

[test_branch_sample_m3_m4.R.zip](https://github.com/braverock/PerformanceAnalytics/files/1194596/test_branch_sample_m3_m4.R.zip)

**Comment by joshuaulrich** [comment]
I rebased the **sample_m3_m4** branch on master, in order to remove the duplicate commits.  The rebased branch is now master, and the tip of the merged branch is now f5172261c54c92c54e6e0e5aa7ce7bda59e3241f.

Closing, since this is now merged to master.

**Pull request closed without merging**